### ISSUE TITLE
Wrong touch handling when centerOffset is used

### ISSUE THREAD

**Issue by marcelofabri** [mapbox/mapbox-gl-native#11829]
**Platform:** iOS
**Mapbox SDK version:** 4.0.0

### Steps to trigger behavior

 1. Apply those changes to the [examples project](https://github.com/mapbox/ios-sdk-examples):

```diff
diff --git a/Examples/Swift/AnnotationViewExample.swift b/Examples/Swift/AnnotationViewExample.swift
index e30240e..2955345 100644
--- a/Examples/Swift/AnnotationViewExample.swift
+++ b/Examples/Swift/AnnotationViewExample.swift
@@ -71,6 +71,20 @@ class AnnotationViewExample_Swift: UIViewController, MGLMapViewDelegate {
 //
 // MGLAnnotationView subclass
 class CustomAnnotationView: MGLAnnotationView {
+    override init(annotation: MGLAnnotation?, reuseIdentifier: String?) {
+        super.init(annotation: annotation, reuseIdentifier: reuseIdentifier)
+        centerOffset = CGVector(dx: 0, dy: -20)
+    }
+
+    override init(reuseIdentifier: String?) {
+        super.init(reuseIdentifier: reuseIdentifier)
+        centerOffset = CGVector(dx: 0, dy: -20)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
     override func layoutSubviews() {
         super.layoutSubviews()
         

```
 2. Run the project and open `AnnotationViewExample_Swift`
 3. Tap on the 20pt area below the annotation (you can use Reveal.app to make it obvious)

### Expected behavior

The touch would be ignored

### Actual behavior

The annotation is selected

---

I'm trying to add pin-like annotations to the map, where you want the tip point to be on the coordinate, (not the center of the annotation). If this is not the recommended way to accomplish this, I'd love to hear other alternatives. Having a sample example that uses `centerOffset` would also be helpful.

**Comment by friedbunny** [comment]
Thanks for the report — is this a regression you’re seeing in 4.0.0? It looks like https://github.com/mapbox/mapbox-gl-native/pull/5931 should have addressed this use case.

**Comment by marcelofabri** [comment]
@friedbunny this is the first time I’m using Mapbox, so I don’t know if it’s a regression

**Comment by marcelofabri** [comment]
Can someone please confirm if this is indeed an issue or if I'm misusing the API? 😬

**Comment by friedbunny** [comment]
@marcelofabri If you’re wanting to add additional tappable space above an annotation view, you would use a positive number — a negative number will _add_ space at the bottom. It’s not currently possible to use this property to subtract from the tap target and tapping the view annotation’s true frame will always select that annotation (https://github.com/mapbox/mapbox-gl-native/pull/5931).

[The docs](https://www.mapbox.com/ios-sdk/api/4.0.0/Classes/MGLAnnotationView.html#/c:objc(cs)MGLAnnotationView(py)centerOffset) are a little unclear about the intention behind this property, so there may be room to improve there.

/cc @frederoni

**Comment by marcelofabri** [comment]
@friedbunny I think the docs are actually wrong:

> Set the offset if the annotation view’s visual center point is somewhere other than the logical center of the view. For example, the view may contain an image that depicts a downward-pointing pushpin or thumbtack, with the tip positioned at the center-bottom of the view. In that case, you would set the offset’s dx to zero and its dy to half the height of the view.

Setting a positive `dy` makes the annotation move down, not up, so pin ends up even further away from the expected position.

For example, here I'm setting `dy` to half the height of the annotation:

![simulator screen shot - iphone 6s - 2018-05-10 at 14 46 02](https://user-images.githubusercontent.com/833072/39896012-6184d95a-5461-11e8-8cd4-fdabc1883b69.png)

This is without setting `centerOffset` at all:

![simulator screen shot - iphone 6s - 2018-05-10 at 14 46 30](https://user-images.githubusercontent.com/833072/39896013-619c8046-5461-11e8-8125-73bdd5617263.png)

This is setting a negative `dy`, which looks right but the tap area is wrong:

![simulator screen shot - iphone 6s - 2018-05-10 at 14 51 41](https://user-images.githubusercontent.com/833072/39896112-b6173774-5461-11e8-9987-92637ba7f3c9.png)


I'm open about alternatives to implement this in other ways. I wanted to avoid having to double the annotation view height and having to deal with this on every annotation view.

**Comment by marcelofabri** [comment]
@friedbunny It looks like the same thing happens on 3.7.5.

**Comment by JamieJeon** [comment]
I have the same problem. Is there any solution?

**Comment by JamieJeon** [comment]
Thank you for sending me the email.
I want to get the answer if there is a solution to this problem.
Could you email back when they come back from vacation?

Thank you.

2018-06-19 11:20 GMT+09:00 Leslie Nakajima <notifications@github.com>:

**Comment by julianrex** [comment]
This is addressed in the upcoming `4.2.0` release.
